src/osd/ReplicatedPG.cc

```diff
@@ -1746,6 +1746,14 @@ void ReplicatedPG::do_op(OpRequestRef& op)
     return;
   }
 
+  // test the ondisk_read_lock as early as possible to avoid stuck at a later time
+  if (op->may_read() && !obc->test_ondisk_read_lock()) {
+    dout(20) << __func__ << " waiting for ondisk_read_lock " << dendl;
+    op->mark_delayed("waiting for ondisk_read_lock");
+    close_op_ctx(ctx, -EBUSY);
+    return;
```

Comment:
Thanks @liewegas for the quick review.

Right this patch does not work. (I thought of using the same machinery as ObjectContext::rwstate did, but apparently rwstate achieved this by its own waiting-list/notify/re-queue mechanism, sorry)

This is actually not an async read problem, the OP we observed having this problem was the bucket_prepare_op/bucket_complete_op, which did a read-modify-write (it has both READ and WRITE flags).

Let me summarize the problem we observed and send an email to ceph-devel for further discussion.